Add Mr-SEQL time series classifier

lnthach opened the pull request:
Hi,

we would like to contribute the implementation of our time series classifier Mr-SEQL as part of the sktime package. 
Mr-SEQL is a univariate time series classifier that learn from multiple symbolic representations of time series (https://link.springer.com/article/10.1007%2Fs10618-019-00633-3).
Our implementation (github.com/lnthach/Mr-SEQL/) was written in C++ hence we provide a Cython wrapper for sktime.

#### What does this implement/fix? Explain your changes.
Add the implementation of Mr-SEQL in shapelet_based directory.
Provide an example for Mr-SEQL in the examples directory.

Best,
Thach

# PR checklist
- [x] I have added unit tests and made sure they pass locally. 
- [x] I have updated the existing example notebooks or provided a new one to showcase how my algorithm works.
- [x] I have updated sktime's [estimator overview](https://github.com/alan-turing-institute/sktime/blob/master/ESTIMATOR_OVERVIEW.md) and I'm committed to maintain my contribution and provide bug fixes if necessary.

fkiraly commented:
Thanks for the contribution, and welcome!

It looks like your PR is stale and you want to re-base before making it, would you mind doing this to avoid merge strangeness?

PS: is your co-author Martin O'Reilly identical with @martintoreilly ?

lnthach commented:
Thanks !
I don't mind at all but I have little experience with Git so I am going to follow this guide 
https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request
Could you please have a look and let me know if this is exactly what you want ?

PS: they are different people. My co-author is based in Dublin.

fkiraly commented:
Looks like:
- my quick diagnosis was wrong, error of sign - you were in fact commits *ahead* instead of *behind*
- this was because you wanted to merge to the wrong branch (master)
- @mloning fixed this by switching the base branch to dev

So, all ready to go now, @mloning?

mloning commented:
@lnthach sorry that it took us so long, we were reviewing some of our contributing guidelines. Your PR looks good to me, but before we merge it, could you please add your method to our [new estimator overview table](https://github.com/alan-turing-institute/sktime/blob/dev/ESTIMATOR_OVERVIEW.md) to keep track of contributions? 

I've also updated your initial comment with a checklist, adding a unit test would be great too!

lnthach commented:
Hi @mloning ,

thank you for getting back. 
I have added the unit test [here](https://github.com/lnthach/sktime/blob/mrseql/sktime/classifiers/shapelet_based/tests/test_mrseql.py) and edit the estimator overview table.  However, I can see the build is failing because of the test although it passed locally. Do you have any suggestion ?

mloning commented:
Here's the error: 
```
from sktime.classifiers.shapelet_based.mrseql import MrSEQLClassifier
1458E   ModuleNotFoundError: No module named 'sktime.classifiers.shapelet_based.mrseql'
```
But the class seems to be in `sktime/classifiers/shapelet_based/mrseql_wrapper.pyx`. 

So, I'd try to change the imports to `from sktime.classifiers.shapelet_based import MrSEQLClassifier`. I'd also create a submodule within `shapelet_based` simply called `mrseql` (rather than `mrseql_cpp`) which contains all of the mrseql code, you can then load it from there into the `shapelet_based.__init__.py` to make it available on the top level.

lnthach commented:
Thanks @mloning .
Unfortunately I haven't been able to resolve the error. I have tried several solutions (including your suggestion) without success. In every attempt, I could build and test locally without any issue. 

`sktime/classifiers/compose/tests/test_columnensembler.py ....                                                                                                                                 [ 11%]
sktime/classifiers/distance_based/tests/test_proximity.py ...                                                                                                                                 [ 11%]
sktime/classifiers/shapelet_based/tests/test_mrseql.py .                                                                                                                                      [ 12%]
sktime/forecasters/tests/test_ARIMA.py ....                                                                                                                                                   [ 12%]`

I have looked around and found some suggestions that it might be because the compiled cython module isn't there when invoking pytest.  Installing in editable mode somehow can avoid this (the import error indeed showed up in the test phase if I installed sktime/mrseql without the editable option). Could you please have a look if this is the cause ?

mloning commented:
I'll look into it when I find some time, but don't have too much experience with debugging Cython or C++ compilation issues. Feel free to play around with the CI files on here to see if you can find out what's going on.

lnthach commented:
Hi @mloning ,

it was actually the init file I added in the test folder that caused the error. I have removed it and the build is successful now. I also have rebased it to make sure it syncs with the latest development. Please let me know if there is anything else I need to do.

Best,
Thach

mloning commented on sktime/classifiers/shapelet_based/mrseql/mrseql.pyx:
```diff
@@ -0,0 +1,439 @@
+
+
+######################### Mr-SEQL (main class) #########################
+''' Time Series Classification with multiple symbolic representations and SEQL (Mr-SEQL)
```
Docstring should go inside the class, I know we haven't been very good at it ourselves, but please try to adhere to scikit-learn's code/docstring formatting.

mloning commented on sktime/classifiers/shapelet_based/mrseql/mrseql.pyx:
```diff
@@ -0,0 +1,439 @@
+        else:
+            self.config = symrepconfig
+
+        #self.label_dict = {} # for translating labels since seql only accept [1.-1] as labels
```
Could you please remove commented code if no longer needed, also it seems as if you should run some auto-formatter, there seem to be a lot of empty lines, etc.

mloning commented on sktime/classifiers/shapelet_based/mrseql/mrseql.pyx:
```diff
@@ -0,0 +1,439 @@
+        full_fm = np.hstack(full_fm)
+        return full_fm
+
+    '''
```
Docstring should go inside of function (see comment above)

mloning commented on sktime/classifiers/shapelet_based/mrseql/mrseql.pyx:
```diff
@@ -0,0 +1,439 @@
+    From dictionary_based/boss.py
+    '''
+    def __X_check(self,X):
+        if isinstance(X, pd.DataFrame):
```
Please also use our input validation methods from [here](https://github.com/alan-turing-institute/sktime/blob/master/sktime/utils/validation/supervised.py), i.e. `validate_X` and `check_is_univariate`

mloning commented on sktime/classifiers/shapelet_based/mrseql/mrseql.pyx:
```diff
@@ -0,0 +1,439 @@
+
+    # map sax features on time series
+    # return ndarray of (number of classes, length of time series)        
+    def map_sax_model(self, ts):
```
Since this is a public function, could you add some docstrings explaining what it does or make it private

mloning commented on sktime/classifiers/shapelet_based/mrseql/mrseql.pyx:
```diff
@@ -0,0 +1,439 @@
+
+
+
+    def summary(self):
```
Same here, please add docstrings or make private

mloning commented on sktime/classifiers/shapelet_based/mrseql/mrseql.pyx:
```diff
@@ -0,0 +1,439 @@
+        self.sfas = {}
+
+
+    def __is_multiclass(self):
```
Please don't use double underscores here and throughout, they have a different meaning in [Python](https://stackoverflow.com/questions/1301346/what-is-the-meaning-of-a-single-and-a-double-underscore-before-an-object-name) and [scikit-learn](https://stackoverflow.com/a/16437327/9334962), single one will be just fine.

lnthach commented:
Hi @mloning 
thank you for your suggestions. I have cleaned the code as suggested and updated the notebook. Please let me know if there are any issues left.
Best,
Thach

mloning commented:
Great! Happy to merge now, anything else from your side before merging?

lnthach commented:
Fantastic. That's all from my side for now. We want to support multivariate time series as well but that's for later.  Thanks a lot @mloning !
